sdk/go/pulumi/provider.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure all outstanding RPCs have completed before proceeding. Also, prevent any new RPCs from happening.
+	pulumiCtx.waitForRPCs()
+	if pulumiCtx.rpcError != nil {
+		return nil, errors.Wrap(pulumiCtx.rpcError, "waiting for RPCs")
+	}
+
+	rpcURN, _, _, err := urn.ToURNOutput().awaitURN(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Serialize all state properties, first by awaiting them, and then marshaling them to the requisite gRPC values.
+	resolvedProps, propertyDeps, _, err := marshalInputs(state)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshaling properties")
+	}
+
+	// Marshal all properties for the RPC call.
+	keepUnknowns := req.GetDryRun()
+	rpcProps, err := plugin.MarshalProperties(
+		resolvedProps,
+		plugin.MarshalOptions{KeepSecrets: true, KeepUnknowns: keepUnknowns, KeepResources: pulumiCtx.keepResources})
+	if err != nil {
+		return nil, errors.Wrap(err, "marshaling properties")
+	}
+
+	// Convert the property dependencies map for RPC and remove duplicates.
+	rpcPropertyDeps := make(map[string]*pulumirpc.ConstructResponse_PropertyDependencies)
+	for k, deps := range propertyDeps {
+		sort.Slice(deps, func(i, j int) bool { return deps[i] < deps[j] })
+
+		urns := make([]string, 0, len(deps))
+		for i, d := range deps {
+			if i > 0 && urns[i-1] == string(d) {
+				continue
+			}
+			urns = append(urns, string(d))
+		}
+
+		rpcPropertyDeps[k] = &pulumirpc.ConstructResponse_PropertyDependencies{
+			Urns: urns,
+		}
+	}
+
+	return &pulumirpc.ConstructResponse{
+		Urn:               string(rpcURN),
+		State:             rpcProps,
+		StateDependencies: rpcPropertyDeps,
+	}, nil
+}
+
+type constructInput struct {
+	value  interface{}
+	secret bool
+	deps   []Resource
+}
+
+// constructInputsMap returns the inputs as a Map.
+func constructInputsMap(inputs map[string]interface{}) Map {
+	result := make(Map, len(inputs))
+	for k, v := range inputs {
+		val := v.(*constructInput)
+		output := newOutput(anyOutputType, val.deps...)
+		output.getState().resolve(val.value, true /*known*/, val.secret, nil)
+		result[k] = output
+	}
+	return result
+}
+
+// constructInputsSetArgs sets the inputs on the given args struct.
+func constructInputsSetArgs(inputs map[string]interface{}, args interface{}) error {
+	if args == nil {
+		return errors.New("args must not be nil")
+	}
+	argsV := reflect.ValueOf(args)
+	typ := argsV.Type()
+	if typ.Kind() != reflect.Ptr || typ.Elem().Kind() != reflect.Struct {
+		return errors.New("args must be a pointer to a struct")
+	}
+	argsV, typ = argsV.Elem(), typ.Elem()
+
+	for k, v := range inputs {
+		val := v.(*constructInput)
+		for i := 0; i < typ.NumField(); i++ {
+			fieldV := argsV.Field(i)
+			if !fieldV.CanSet() {
+				continue
+			}
+			field := typ.Field(i)
+			tag, has := field.Tag.Lookup("pulumi")
+			if !has || tag != k {
+				continue
+			}
+
+			if !field.Type.Implements(reflect.TypeOf((*Input)(nil)).Elem()) {
```

Comment:
[I just found these two comments in drafts that I had intended to send a couple weeks ago!  I believe they are still relevant to the design here though - so sharing now ahead of a more thorough review of the updated PR shortly]

> I didn't want to force component authors to have to define it twice manually like this. But curious if you think we should force component authors to define these the same way as the Go codegen does.

Two things that I've gotten tripped up on myself working with this:
1. How do I write my own nested type to use here?  Like if I want a `Routes: []Route`, how do I write `[]Route` so that it works?  It appears I need to create a `type RouteArrayInput`, ensure that has a `ToRouteArrayOutputWithContext`, and then also a custom `Output` struct that wraps the raw `Route` struct?  I haven't yet quite figured out what works fully here.
2. If we had just alllowed using the same thing that codegen produces, then I could just use the types that the Go SDK codegen already generates for me - by having the provider actually depend on the SDK.  That may be a terrible idea, but it is nice in that the implementation gets to work with the exact same structs/interfaces as consumers will.  It also means I don't have to recreate all of this manually inside the implementation - since there is already a codegen process to create these for the SDK.   However, the types implemented in the SDK don't seem to work correctly with the logic as implemented here.

Reply:
It's also a little strange that these accept Input types - given that at least currently they will always be Outputs, and even once prompt values are possible, the user will always want to be clear here in the types which ones are prompt and which are Output. It will never be desirable for the implementor to see Input types vs. Output types. The implementation will always have to immediately turn the inputs back into outputs to work with them.

If this args struct used Outputs, the implementation here would also be much simpler and less "weird" in that it wouldn't have to discover To<>OutputWithContext to figure out what the output type was to create.

Reply:
> How do I write my own nested type to use here?

The idea was to follow our documented pattern for inputs:

https://github.com/pulumi/pulumi/blob/afe109621e878aaf717044bf9c7f0958c53f6ffa/sdk/go/pulumi/types.go#L767-L829

But it is an unfortunate amount of boilerplate to have to write manually (especially so if you'd need to also manually write associated boilerplate for array variants, map variants, etc.), so I fully empathize with the desire from (2) to use the types that the Go SDK codegen generates. However, ideally, we'd have a way to write a much simpler type in the implementation without all the extra ceremony/boilerplate, and drive the schema + SDK generation based on it, not the other way around.

> However, the types implemented in the SDK don't seem to work correctly with the logic as implemented here.

Hmmm, do you happen to recall the specific issue? I believe there is a bug in the current implementation in this PR, where the unmarshalled value from the gRPC call isn't converted into the struct type -- so it'll remain a map. I am addressing this issue with the changes for prompt values.

> It's also a little strange that these accept Input types - given that at least currently they will always be Outputs, and even once prompt values are possible, the user will always want to be clear here in the types which ones are prompt and which are Output. It will never be desirable for the implementor to see Input types vs. Output types. The implementation will always have to immediately turn the inputs back into outputs to work with them.
>
> If this args struct used Outputs, the implementation here would also be much simpler and less "weird" in that it wouldn't have to discover To<>OutputWithContext to figure out what the output type was to create.

This is a good point. I was thinking you'd write the component implementation the same way you'd write it if you were releasing it as a regular Go library, where (ignoring prompt values for a moment) you would want to type inputs as the `Input` type to support either `Input` or `Output` values (and inside the implementation you'd immediately convert to `Output` to work with it).

But for the implementation of multi-lang components in Go, I could see how typing it as `Output` would simplify things here. We'd immediately know the `Output` type to use and it would save the user from having to convert the value to `Output` inside the component implementation.

I do still feel like we could support `Input` though... I could see supporting either `Output` or `Input`. If it's `Output`, great -- use it. If it's `Input`, one thing I could do is get its `ElementType()` and then see if the output type has been registered with `pulumi.RegisterOutputType` (using it if so), and finally falling back to looking for a `To<>Output` method.

> discover To<>OutputWithContext

Note: after the recent discussion of possibly deprecating the `<>WithContext` methods, I've switched this to look for just `To<>Output`.